finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Stream.scala

```diff
@@ -33,6 +49,12 @@ trait Stream {
    */
   def onEnd: Future[Unit]
 
+  /**
+   * Satisfied when the consumer of a stream cancels it by calling
+   * `cancel()`.
+   */
+  def onCancel: Future[Reset]
```

Comment:
Can this fail?  Does cancelling this future do anything?

If "no" to both of these, consider using an interface like `def onCancel(f: Reset => Unit)` and then have impls like `override def onCancel(f: Reset => Unit) { cancelP.onSuccess(f) }`.

This would have the benefit of hiding some of the complicated Future semantics from this part of the interface if they're not necessary.

Reply:
Also, if we'd ever want to join/select over onCancels, the callback approach would be suboptimal.  worth asking the question though

Reply:
good suggestion, thanks